mlflow/store/artifact/artifact_repo.py

```diff
@@ -330,7 +339,30 @@ def upload_trace_data(self, trace_data: str) -> None:
         Args:
             trace_data: The json-serialized trace data to upload.
         """
-        raise NotImplementedError
+        with write_local_temp_trace_data_file(trace_data) as temp_file:
+            self.log_artifact(temp_file)
+
+
+@contextmanager
+def write_local_temp_trace_data_file(trace_data: str):
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_file = Path(temp_dir, TRACE_DATA_FILE_NAME)
+        with temp_file.open("w") as f:
+            f.write(trace_data)
+        yield temp_file
+
+
+def try_read_trace_data(trace_data_path):
+    if not os.path.exists(trace_data_path):
+        raise MlflowTraceDataNotFound(artifact_path=trace_data_path)
+    with open(trace_data_path) as f:
+        data = f.read()
+    if not data:
+        raise MlflowException("Trace data is empty")
```

Comment:
Do we need a different exception for empty trace data?

Reply:
https://github.com/mlflow/mlflow/pull/11958#discussion_r1596117463